### ISSUE TITLE
How to use semantic segmentation tokenizer for precomputing tokens for this modality? 

### ISSUE THREAD

**Issue by HITESH2002-JAIN** [apple/ml-4m#24]
I am working on precomputing tokens for each modality in my 4M training pipeline. I’m using grayscale semantic segmentation masks as input, but I’m encountering an issue where the regenerated output does not match the original mask.

<img width="534" alt="Screenshot 2024-08-12 at 4 45 24 PM" src="https://github.com/user-attachments/assets/5ca0ac00-ebcf-4a3a-8c57-23ab8f260b71">

This is the code I am using for precomputing tokens
```
from fourm.vq.vqvae import VQVAE, DiVAE
from PIL import Image
from torchvision import transforms
from fourm.utils import denormalize, IMAGENET_INCEPTION_MEAN, IMAGENET_INCEPTION_STD
from torchvision.transforms import Normalize
transform = transforms.ToTensor()
resize = transforms.Resize((224, 224))

tok = VQVAE.from_pretrained('EPFL-VILAB/4M_tokenizers_semseg_4k_224-448').cuda()
tensors_b3hw = []
for image_path in selected_images :
    image = Image.open(image_path)
    rgb_b3hw = transform(resize(image)).unsqueeze(0)  
    tensors_b3hw.append(rgb_b3hw)
stacked_tensors_b3hw = torch.cat(tensors_b3hw, dim=0).int()
squeezed_tensor = torch.squeeze(stacked_tensors_b3hw, dim=1)
squeezed_tensor.shape
_, _, tokens = tok.encode(squeezed_tensor.cuda())
image_size = rgb_b3hw.shape[-1]
output_rgb_b3hw  = tok.decode_tokens(tokens, image_size=image_size)
```
The output_rgb_b3hw tensor, which is the regenerated output, consists of 134 channels. However, this does not match the original mask that I passed to the tokenizer. I expected the output to have the same number of channels as the input mask. 
Am I missing something in the preprocessing or tokenization process? Is there a step I need to adjust to ensure that the regenerated output matches the original mask in terms of channel dimensions?

Any guidance or suggestions would be appreciated. Thank you!

@garjania Please help me out. Am I doing something wrong here ?

**Comment by garjania** [comment]
Hi @HITESH2002-JAIN, what is the dimension of the semseg mask you're using? Does it have the same classes as COCO? Our current semseg tokenizer only supports inputs with the same class labels as COCO. Also, I suggest you use the [SemsegTransform](https://github.com/apple/ml-4m/blob/ee81e40983616e9737f444068572f0bac9961abe/fourm/data/modality_transforms.py#L327) to load and transform your semseg masks.

**Comment by adricarda** [comment]
Hi, I am experinecing a similar issue. Could you please share the exact Mask2former model you used to get pseudo labels?
In the link you provide (https://github.com/facebookresearch/Mask2Former/blob/main/MODEL_ZOO.md) I don't see any model trained on the COCO dataset for semantic segmentation. But all the other model won't be compatible with the tokenizer from my understanding.
Thnak you!

**Comment by garjania** [comment]
Hi @adricarda,

You can use the pretrained weights listed in the [COCO Panoptic Segmentation Model Zoo](https://github.com/facebookresearch/Mask2Former/blob/main/MODEL_ZOO.md#panoptic-segmentation) to generate semantic segmentation maps. Technically, any of the listed panoptic segmentation weights should work, but I suggest using the model with the Swin-B backbone (`model id: 48558700_5`).

**Comment by adricarda** [comment]
Hi, I tried the panoptic model and indeed it outputs segmentation masks for both "stuff" and "things", hence the 133 classes.
Thank you!
